--- a/dwm.c	2011-07-10 22:12:16.720000000 +0300
+++ b/dwm.c	2011-07-10 22:15:17.964000003 +0300
@@ -880,15 +880,17 @@ enternotify(XEvent *e) {
 
 	if((ev->mode != NotifyNormal || ev->detail == NotifyInferior) && ev->window != root)
 		return;
-	c = wintoclient(ev->window);
-	m = c ? c->mon : wintomon(ev->window);
-	if(m != selmon) {
-		unfocus(selmon->sel, True);
-		selmon = m;
+	if(!focusonclick) {
+		c = wintoclient(ev->window);
+		m = c ? c->mon : wintomon(ev->window);
+		if(m != selmon) {
+			unfocus(selmon->sel, True);
+			selmon = m;
+		}
+		else if(!c || c == selmon->sel)
+			return;
+		focus(c);
 	}
-	else if(!c || c == selmon->sel)
-		return;
-	focus(c);
 }
 
 void
