--- a/dwm.c	2011-07-10 23:39:43.491000025 +0300
+++ b/dwm.c	2011-07-10 23:39:46.343000023 +0300
@@ -263,6 +263,7 @@ static void pushdown(const Arg *arg);
 static void cycle(const Arg *arg);
 static int shifttag(int dist);
 static void tagcycle(const Arg *arg);
+static void gaplessgrid(Monitor *m);
 
 /* variables */
 static const char broken[] = "broken";
@@ -2237,6 +2238,43 @@ tagcycle(const Arg *arg) {
 	view(&a);
 }
 
+void
+gaplessgrid(Monitor *m) {
+	unsigned int n, cols, rows, cn, rn, i, cx, cy, cw, ch;
+	Client *c;
+
+	for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next))
+		n++;
+	if(n == 0)
+		return;
+
+	/* grid dimensions */
+	for(rows = 0; rows <= n/2; rows++)
+		if(rows*rows >= n)
+			break;
+	if(n == 5) /* set layout against the general calculation: not 1:2:2, but 2:3 */
+		rows = 2;
+	cols = n/rows;
+
+	/* window geometries */
+	ch = rows ? m->wh / rows : m->wh;
+	cn = 0; /* current column number */
+	rn = 0; /* current row number */
+	for(i = 0, c = nexttiled(m->clients); c; i++, c = nexttiled(c->next)) {
+		if(i/cols + 1 > rows - n%rows)
+			cols = n/rows + 1;
+		cw = cols ? m->ww / cols : m->ww;
+		cx = m->wx + cn*cw;
+		cy = m->wy + rn*ch;
+		resize(c, cx, cy, cw - 2 * c->bw, ch - 2 * c->bw, False);
+		cn++;
+		if(cn >= cols) {
+			cn = 0;
+			rn++;
+		}
+	}
+}
+
 int
 main(int argc, char *argv[]) {
 	if(argc == 2 && !strcmp("-v", argv[1]))
